### PR TITLE
Add gelman_rubin and effective_n to df_summary (#2532)

### PR DESCRIPTION
**Pull request by SioKCronin**
Hi there,

I'm submitting this [WIP] PR to ensure what I'm building is consistent with the spirit of `stats.py` and `diagnostics.py`. 

The approach I'm taking to add`effective_n` and `gelman_rubin` to `df_summary` in `stats.py` is to add them to the list of `funcs` that are iterated over by `df_summary`. However, this will require functions that take specific var values, as opposed to accepting a trace, as is the case in the `diagnostics.py` implementation. 

My question is whether it makes more sense to have these new "atomic" functions (the ones that process var values directly) in `stats.py` where they can then be used by `df_summary` and then imported by `diagonistics.py` to be rolled up into a wrapper function that iterates over `varnames`, or whether the functions in `diagnostics.py` should be split out into an atomic portion (that can be imported into `stats.py`) and a wrapper that takes traces. My hunch is to have the atomic functions in `stats.py`, yet would love feedback. 

Also, two quick questions that I imagine will be obvious to others: 

1. When I get `RuntimeWarning: Degrees of freedom <= 0 for slice`, does that relate to the shape of the test input generated by ndarray.py? 

2. Since in `effective_n` in diagnostics.py it appears `get_vhat(x)` returns a scaler (`Vhat = get_vhat(x)`), how can it be indexed (i.e. ` _n_eff[tup] = get_neff(x[tup], Vhat[tup])`)?. I'm including `get_vhat(x)` and a portion of `effective_n` from diagnostics.py for context:

```
def get_vhat(x):
    # number of chains is last dim (-1)
    # chain samples are second to last dim (-2)
    num_samples = x.shape[-2]

    # Calculate between-chain variance
    B = num_samples * np.var(np.mean(x, axis=-2), axis=-1, ddof=1)

    # Calculate within-chain variance
    W = np.mean(np.var(x, axis=-2, ddof=1), axis=-1)

    # Estimate of marginal posterior variance
    Vhat = W * (num_samples - 1) / num_samples + B / num_samples

    return Vhat
```
```
for var in varnames:
    x = np.array(mtrace.get_values(var, combine=False))

    # make sure to handle scalars correctly - add extra dim if needed
    if len(x.shape) == 2:
        is_scalar = True
        x = np.atleast_3d(mtrace.get_values(var, combine=False))
    else:
        is_scalar = False

    # now we are going to transpose all dims - makes the loop below
    # easier by moving the axes of the variable to the front instead
    # of the chain and sample axes
    x = x.transpose()

    Vhat = get_vhat(x)

    # get an array the same shape as the var
    _n_eff = np.zeros(x.shape[:-2])

    # iterate over tuples of indices of the shape of var
    for tup in np.ndindex(*list(x.shape[:-2])):
        _n_eff[tup] = get_neff(x[tup], Vhat[tup])
```

### CODE REVIEW

**Comment by ColCarroll** [comment]
I haven't looked much at this code, but from an architecture view, functions that (try to) measure the convergence of a trace should go into `diagnostics.py`.  

That said, I think we should favor vectorized functions over non-vectorized functions, so maybe the atomic ones might not be necessary?  Do you think it would be possible to replace the `funcs` array with a dictionary, something like 
```
funcs = {
    'vectorized': {
        'mean': lambda x: np.mean(x, axis=1),
        'std': lambda x: np.std(x, axis=0),
        'Rhat': lambda x: gelman_rubin(x),
        ...
    },
    'atomic': [
        anything_left_and_stat_funcs,
        ...
    ]
}

```

Then all the vectorized functions would be _much_ faster.  I haven't looked closely, so this might be a silly suggestion!

**Comment by SioKCronin** [comment]
Thanks for the clarity on `diagnostics.py`! @ColCarroll 📊 

Also, thanks for the vectorization suggestion. I believe that should be possible here. 

My hope was to simply import `effective_n` and `gelman_rubin` from `diagnostics.py`, and do a little magic to get them in line with the other funcs, but since the procedure is set up to take arrays instead of traces, I think I'll include a paired down copy of the functions in `stats.py` and handle them separately when they're called.

**Comment by twiecki** [comment]
I don't think the functions should be copied. Rather, the original functions should be changed to accept arrays or traces.

**Comment by SioKCronin** [comment]
Thanks @twiecki --- I'll proceed in that direction.

**Comment by junpenglao** [comment]
Any update @SioKCronin?

**Comment by SioKCronin** [comment]
I'm so sorry for my delay! I just started grad school, but am now a few weeks in and getting settled.

I'm working on this today, and think I just need to check the type (array vs. mtrace) on the stats themselves, and proceed from there.

**Comment by SioKCronin** [comment]
Hi @junpenglao ! I've updated `gelman_rubin` and `effective_n` in diagnostics.py so that each now accepts either a MultiTrace or a trace object (returning a dict of floats for MultiTrace or a float for a trace object). They are now both included in the list of funcs in `summary_df` in stats.py. 

I need to build out some test coverage for this new input to these stats, and will acquaint myself with examples in the test suite to make sure I'm consistent. 

I also need to update `df_summary` documentation to include sample output for both `gelman_rubin` and `effective_n`, but I think that is just output generated from one of the test traces?

File: pymc3/stats.py [hunk 1]
```diff
@@ -354,10 +343,10 @@ def bpic(trace, model=None):
 
 def compare(traces, models, ic='WAIC', method='stacking', b_samples=1000,
             alpha=1, seed=None, round_to=2):
-    R"""Compare models based on the widely available information criterion (WAIC)
+    """Compare models based on the widely available information criterion (WAIC)
```
Comment (twiecki) [inline pymc3/stats.py]: These need `R`.

Comment (SioKCronin) [reply pymc3/stats.py]: Thanks!

File: pymc3/diagnostics.py [hunk 2]
```diff
@@ -141,6 +142,22 @@ def gelman_rubin(mtrace, varnames=None, include_transformed=False):
...
+
+        # Calculate within-chain variance
+        W = np.mean(np.var(x, axis=1, ddof=1), axis=0)
+
+        # Estimate of marginal posterior variance
+        Vhat = W * (num_samples - 1) / num_samples + B / num_samples
+
+        return np.sqrt(Vhat / W)
+
+    if not isinstance(mtrace,MultiTrace):
```
Comment (junpenglao) [inline pymc3/diagnostics.py]: Need space here `(mtrace, MultiTrace)` (PEP8)

File: pymc3/stats.py [hunk 3]
```diff
@@ -795,6 +782,10 @@ def df_summary(trace, varnames=None, transform=lambda x: x, stat_funcs=None,
         mu__0  0.106897  0.066473  0.001818 -0.020612  0.231626
         mu__1 -0.046597  0.067513  0.002048 -0.174753  0.081924
 
+                  n_eff      Rhat
+        mu__0       6.0  7.509734
```
Comment (junpenglao) [inline pymc3/stats.py]: here you should change the n_eff and Rhat to a more reasonable number, eg n_eff=487.0 Rhat=1.00001

File: pymc3/stats.py [hunk 4]
```diff
@@ -104,7 +103,7 @@ def autocov(x, lag=1):
 
 def dic(trace, model=None):
     """Calculate the deviance information criterion of the samples in trace from model
-    Read more theory here - in a paper by some of the leading authorities on model selection -
+    Read more theory here - in a paper by some of the leading authorities on Model Selection -
```
Comment (junpenglao) [inline pymc3/stats.py]: Could you please change this back (also some of the changes below)? Some of the change in this file is related to https://github.com/pymc-devs/pymc3/pull/2606, and you are changing them back here (accident during merging master to your branch?). If it helps you can also copy from master and just reimplement you change.

**Review by junpenglao — CHANGES_REQUESTED** [review]

File: pymc3/stats.py [hunk 5]
```diff
@@ -724,10 +712,9 @@ def quantiles(x, qlist=(2.5, 25, 50, 75, 97.5), transform=lambda x: x):
         return dict(zip(qlist, quants))
 
     except IndexError:
-        pm._log.warning("Too few elements for quantile calculation")
+        _log.warning("Too few elements for quantile calculation")
```
Comment (junpenglao) [inline pymc3/stats.py]: since you deleted `import pymc3 as pm` above, here will cause error.

File: pymc3/diagnostics.py (hunk 2 above)
Comment (SioKCronin) [reply pymc3/diagnostics.py]: Thank you.

File: pymc3/stats.py [hunk 6]
```diff
@@ -160,14 +160,12 @@ def logp_vals_point(pt):
                 logp = logp[~var.observations.mask]
             logp_vals.append(logp.ravel())
 
-        return np.concatenate(logp_vals)
```
Comment (junpenglao) [inline pymc3/stats.py]: You should not delete this.

File: pymc3/stats.py [hunk 7]
```diff
@@ -160,14 +160,12 @@ def logp_vals_point(pt):
...
-
     try:
         points = trace.points()
     except AttributeError:
         points = trace
 
-    points = tqdm(points) if progressbar else points
```
Comment (junpenglao) [inline pymc3/stats.py]: You should not change this.

File: pymc3/stats.py [hunk 8]
```diff
@@ -833,7 +841,6 @@ def summary(trace, varnames=None, transform=lambda x: x, stat_funcs=None,
         var_df = pd.concat([f(flat_vals) for f in stat_funcs], axis=1)
         var_df.index = ttab.create_flat_names(var, vals.shape[1:])
         var_dfs.append(var_df)
-    return pd.concat(var_dfs, axis=0)
```
Comment (junpenglao) [inline pymc3/stats.py]: You should not delete this.

File: pymc3/stats.py [hunk 9]
```diff
@@ -1012,7 +1018,8 @@ def _groupby_leading_idxs(shape):
 
 
 def bfmi(trace):
-    R"""Calculate the estimated Bayesian fraction of missing information (BFMI).
```
Comment (junpenglao) [inline pymc3/stats.py]: Unnecessary change (same the one below).

**Review by junpenglao — CHANGES_REQUESTED** [review]
OK, looks good to me overall. Could you make a few more changes?

File: pymc3/stats.py (hunk 9 above)
Comment (SioKCronin) [reply pymc3/stats.py]: Thanks @junpenglao ! I'm not sure what happened with my upstream master merging, but I'll make sure to look into that before starting any further PR.

**Review by junpenglao — APPROVED** [review]

**Comment by junpenglao** [comment]
Awesome! Thanks for the contribution!

**Comment by SioKCronin** [comment]
It was super fun, and I'm glad I understand the codebase better now.
